flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,734 @@
+        } catch (Exception e) {
+            e.printStackTrace();
+            fail(e.getMessage());
+        }
+    }
+
+    private void waitAndAssert(StreamGraph streamGraph) {
+        waitUntilJobFinished(streamGraph);
+        assertEquals(CollectionSink.getExpectedResult(), PreHandlingSource.getActualResult());
+    }
+
+    /** A source consumes data which is generated randomly and supports pre-handling record. */
+    private static class PreHandlingSource extends RichSourceFunction<Integer>
+            implements CheckpointedFunction, CheckpointListener {
+
+        protected volatile int currentIndex;
+
+        protected final AtomicInteger completedCheckpointNum;
+
+        protected volatile boolean isCanceled;
+
+        private static List<Integer> sourceList;
+
+        private transient ListState<Integer> lastSentStored;
+
+        public PreHandlingSource() {
+            this.completedCheckpointNum = new AtomicInteger();
+        }
+
+        public static void initSourceData(int totalNum) {
+            sourceList = new ArrayList<>(totalNum);
+            for (int i = 0; i < totalNum; i++) {
+                sourceList.add(ThreadLocalRandom.current().nextInt(totalNum));
+            }
+        }
+
+        public static Map<Integer, Integer> getActualResult() {
+            return sourceList.stream()
+                    .collect(
+                            Collectors.toConcurrentMap(
+                                    Function.identity(), element -> 1, Integer::sum));
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            lastSentStored.update(Arrays.asList(currentIndex, completedCheckpointNum.get()));
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            lastSentStored =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("counter", Integer.class));
+            if (context.isRestored()) {
+                Iterable<Integer> indexAndCheckpointNum = lastSentStored.get();
+                currentIndex = get(indexAndCheckpointNum, 0);
+                completedCheckpointNum.compareAndSet(0, get(indexAndCheckpointNum, 1));
+            }
+        }
+
+        @Override
+        public void run(SourceContext<Integer> ctx) throws Exception {
+            while (!isCanceled && currentIndex < sourceList.size()) {
+                preHandle();
+                synchronized (ctx.getCheckpointLock()) {
+                    if (!isCanceled && currentIndex < sourceList.size()) {
+                        int currentElement = sourceList.get(currentIndex++);
+                        ctx.collect(currentElement);
+                    }
+                }
+            }
+        }
+
+        protected void preHandle() throws Exception {
+            // do nothing by default
```

Comment:
How about renaming to `beforeElement` or `beforeEmitElement`?
I'd also add javadoc saying that it's run without holding the checkpoint lock.